src/hotspot/cpu/x86/stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,2012 @@
+  __ addq(hsPtrRet, 32);
+
+  __ cmpq(hsPtrRet, last);
+  __ cmovq(Assembler::aboveEqual, hsPtrRet, last);
+
+  __ bind(temp);
+
+  // compare_big_haystack_to_needle will jump to loop_top until a match has been
+  // found
+  compare_big_haystack_to_needle(sizeKnown, size, NUMBER_OF_NEEDLE_BYTES_TO_COMPARE, loop_top, hsPtrRet, hsLength,
+                                 needleLen, isU, DO_EARLY_BAILOUT, eq_mask, temp2, r10, _masm);
+
+  // At this point, we have at least one "match" where first and last bytes
+  // of the needle are found the correct distance apart.
+  //
+  // NOTE: haystack (rbx) should be preserved; hsPtrRet(rcx) is expected to
+  //    point to the haystack such that hsPtrRet[tzcntl(eq_mask)] points to
+  //    the matched string.
+}
+
+static void preload_needle_helper(int size, Register needle, Register needleVal, StrIntrinsicNode::ArgEncoding ae,
+                                  MacroAssembler *_masm) {
+  // Pre-load the value (correctly sized) of the needle for comparison purposes.
+
+  assert_different_registers(needle, needleVal);
+
+  bool isLL = (ae == StrIntrinsicNode::LL);
+  bool isUL = (ae == StrIntrinsicNode::UL);
+  bool isUU = (ae == StrIntrinsicNode::UU);
+  bool isU = isUL || isUU;  // At least one is UTF-16
+
+  int bytesAlreadyCompared = 0;
+  int bytesLeftToCompare = 0;
+  int offsetOfFirstByteToCompare = 0;
+
+  bytesAlreadyCompared = isU ? (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE * 2) : NUMBER_OF_NEEDLE_BYTES_TO_COMPARE;
+  offsetOfFirstByteToCompare
+      = isU ? ((NUMBER_OF_NEEDLE_BYTES_TO_COMPARE - 1) * 2) : (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE - 1);
+
+  bytesLeftToCompare = size - bytesAlreadyCompared;
+  assert((bytesLeftToCompare <= 8), "Too many bytes left to compare");
+
+  if (bytesLeftToCompare <= 0) {
+    return;
+  }
+
+  // At this point, there is at least one byte of the needle that needs to be
+  // compared to the haystack.
+
+  if (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE <= 2) {
+    switch (bytesLeftToCompare) {
+    case 1:
+    case 2:
+      __ movzwl(needleVal, Address(needle, offsetOfFirstByteToCompare));
+      break;
+
+    case 3:
+    case 4:
+      __ movl(needleVal, Address(needle, offsetOfFirstByteToCompare));
+      break;
+
+    case 5:
+      // Read one byte before start of needle, then mask it off
+      __ movq(needleVal, Address(needle, (offsetOfFirstByteToCompare - 2)));
+      __ shrq(needleVal, 0x8);
+      break;
+
+    case 6:
+      __ movq(needleVal, Address(needle, (offsetOfFirstByteToCompare - 1)));
+      break;
+
+    case 7:
+    case 8:
+      __ movq(needleVal, Address(needle, offsetOfFirstByteToCompare));
+      break;
+
+    default:
+      break;
+    }
+  }
+
+  if (NUMBER_OF_NEEDLE_BYTES_TO_COMPARE >= 3) {
+    switch (bytesLeftToCompare) {
+    case 1:
+    case 2:
+      __ movl(needleVal, Address(needle, (offsetOfFirstByteToCompare - 2)));
+      break;
+
+    case 3:
+    case 4:
+      __ movl(needleVal, Address(needle, offsetOfFirstByteToCompare));
```

Comment:
If the size of the needle is 7 and it is an LL case with NUMBER_OF_NEEDLE_BYTES_TO_COMPARE set as 3:
  bytesLeftToCompare = 4 (i.e. 7-3);
  offsetOfFirstByteToCompare = 2 (i.e. 3-1);
  the movl will be loading bytes 2,3,4,5
So we seem to be missing loading the last byte of the needle. Is that correct?

Reply:
Bytes 0, 1, and 6 have already compared equal before getting to this code, so it is correct functionally.